Mathlib/Algebra/Opposites.lean

```diff
@@ -0,0 +1,382 @@
+canonical map from `α` to `αᵐᵒᵖ`.
+
+We also define `AddOpposite α = αᵃᵒᵖ` to be the additive opposite of `α`. It inherits all
+multiplicative algebraic structures on `α` (in other files), and reverses the order of summands in
+additive structures, i.e. `op (x + y) = op y + op x`, where `AddOpposite.op` is the canonical map
+from `α` to `αᵃᵒᵖ`.
+
+## Notation
+
+* `αᵐᵒᵖ = MulOpposite α`
+* `αᵃᵒᵖ = AddOpposite α`
+
+## Tags
+
+multiplicative opposite, additive opposite
+-/
+
+
+universe u v
+
+open Function
+
+/-- Multiplicative opposite of a type. This type inherits all additive structures on `α` and
+reverses left and right in multiplication.-/
+@[to_additive
+      "Additive opposite of a type. This type inherits all multiplicative structures on
+      `α` and reverses left and right in addition."]
+def MulOpposite (α : Type u) : Type u :=
+  α
+#align mul_opposite MulOpposite
+
+/-- Multiplicative opposite of a type. -/
+postfix:max "ᵐᵒᵖ" => MulOpposite
+
+/-- Additive opposite of a type. -/
+postfix:max "ᵃᵒᵖ" => AddOpposite
+
+namespace MulOpposite
+
+/-- The element of `MulOpposite α` that represents `x : α`. -/
+-- porting note: the attribute `pp_nodot` does not exist yet
+--@[pp_nodot,
+@[to_additive "The element of `αᵃᵒᵖ` that represents `x : α`."]
+def op : α → αᵐᵒᵖ :=
+  id
+#align mul_opposite.op MulOpposite.op
+
+/-- The element of `α` represented by `x : αᵐᵒᵖ`. -/
+--@[pp_nodot,
+@[to_additive "The element of `α` represented by `x : αᵃᵒᵖ`."]
+def unop : αᵐᵒᵖ → α :=
+  id
+#align mul_opposite.unop MulOpposite.unop
+
+-- attribute [pp_nodot] AddOpposite.op AddOpposite.unop
+
+@[simp, to_additive]
+theorem unop_op (x : α) : unop (op x) = x :=
+  rfl
+#align mul_opposite.unop_op MulOpposite.unop_op
+
+@[simp, to_additive]
+theorem op_unop (x : αᵐᵒᵖ) : op (unop x) = x :=
+  rfl
+#align mul_opposite.op_unop MulOpposite.op_unop
+
+@[simp, to_additive]
+theorem op_comp_unop : (op : α → αᵐᵒᵖ) ∘ unop = id :=
+  rfl
+#align mul_opposite.op_comp_unop MulOpposite.op_comp_unop
+
+@[simp, to_additive]
+theorem unop_comp_op : (unop : αᵐᵒᵖ → α) ∘ op = id :=
+  rfl
+#align mul_opposite.unop_comp_op MulOpposite.unop_comp_op
+
+/-- A recursor for `MulOpposite`. Use as `induction x using MulOpposite.rec`. -/
+@[simp, to_additive "A recursor for `AddOpposite`. Use as `induction x using AddOpposite.rec`."]
+protected def rec {F : ∀ _ : αᵐᵒᵖ, Sort v} (h : ∀ X, F (op X)) : ∀ X, F X := fun X => h (unop X)
+#align mul_opposite.rec MulOpposite.rec
+
+/-- The canonical bijection between `α` and `αᵐᵒᵖ`. -/
+@[to_additive "The canonical bijection between `α` and `αᵃᵒᵖ`.",--]
+  simps (config := { fullyApplied := false }) apply]
+def opEquiv : α ≃ αᵐᵒᵖ :=
+  ⟨op, unop, unop_op, op_unop⟩
+#align mul_opposite.op_equiv MulOpposite.opEquiv
+
+-- porting note: We have to provide this lemma by hand, since the autogenerated version makes the
+-- simpNF linter unhappy
+lemma opEquiv_symm_apply : (opEquiv.symm.toFun : αᵐᵒᵖ → α) = unop := rfl
```

Comment:
Shouldn't this be written without the toFun? The lemma should be about the coercion.

Reply:
Isn't that the same thing? My understanding was that `(opEquiv.symm : αᵐᵒᵖ → α)` and `(opEquiv.symm.toFun : αᵐᵒᵖ → α)` are syntactically equal.

Reply:
but a good point here is that the recent changes in the `funLike.coe` makes it so that the `simps` does the right thing again